### PR TITLE
Reality tile reprojection

### PR DESCRIPTION
**Pull request by RBBentley**
Add reprojection from reality tile  to iModel coordinates through the GCS in a similar manner to map tile reprojection.   This is not required for typical reality model attachments as a linear transform is sufficiently accurate unless both the iModel and reality model are spatially large.  This is currently required when the Cesium OpenStreetMap building dataset is attached to very large iModels and the curvature of the earth can errors in altitude.  The usage of reprojection is currently limited to tiles that have boundaries defined by latitude/longitude regions (currently only the Cesium OSM) dataset to avoid unnecessary and detrimental reprojection of large, but essentially rectilinear tilesets (point clouds).

### CODE REVIEW

**Review by pmconne — CHANGES_REQUESTED** [review]
rush change, extract-api

**Comment by pmconne** [comment]
Have you queued an ImageTests run?

**Comment by pmconne** [comment]
Think @AlainRobertAtBentley ought to review?

**Review by paulius-valiunas — APPROVED** [review]
Approving for whitespace changes in eslint plugin

**Review by RBBentley — COMMENTED** [review]
> Have you queued an ImageTests run?

Yes.... in progress.  I'll add more test cases for the OSM buildings as well.

**Comment by RBBentley** [comment]
> rush change, extract-api

Done

**Comment by RBBentley** [comment]
> Have you queued an ImageTests run?

Only difference was correction of the building heights in Philadelphia (fixed by this PR)
image

**Comment by RBBentley** [comment]
> Think someone like @AlainRobertAtBentley ought to review?

Added Alain as reviewer

File: core/frontend/src/tile/RealityTileTree.ts [hunk 1]
```diff
@@ -104,26 +106,53 @@ export class TraversalSelectionContext {
...
     super(params);
     this.loader = params.loader;
     this.yAxisUp = true === params.yAxisUp;
     this._rootTile = this.createTile(params.rootTile);
+    this.cartesianRange = BackgroundMapGeometry.getCartesianRange(this.iModel);
+    this.cartesianTransitionDistance = this.cartesianRange.diagonal().magnitudeXY() * .25;      // Transition distance from elliptical to cartesian.
+    this._gcsConverter = params.gcsConverterAvailable ? params.iModel.geoServices.getConverter("WGS84") : undefined;
+    if (params.rootToEcef) {
+      this._rootToEcef = params.rootToEcef;
+      const dbToRoot = this.iModelTransform.inverse()!;
```
Comment (pmconne) [inline core/frontend/src/tile/RealityTileTree.ts:152]: Handle non-invertible transform?

Comment (RBBentley) [reply core/frontend/src/tile/RealityTileTree.ts]: Done...

File: core/frontend/src/tile/RealityTile.ts [hunk 2]
```diff
@@ -275,4 +311,124 @@ export class RealityTile extends Tile {
...
+  }
+  public override disposeContents(): void {
+    super.disposeContents();
+    this._reprojectedGraphic = dispose(this._reprojectedGraphic);
+  }
+}
+
+/** When additive refinement is used (as in the Cesium OSM tileset) it is not possible to accurately reproject very large, low level tiles
+ * In this case we create addtitional "step" children (grandchildren etc. ) that will clipped portions display the their ancestor's additive geometry.
+ * These step children are subdivided until they are small enough to be accurately reprojected - this is controlled by the additiveRefinementThreshold (currently 20KM).
```
[the remark this answers is not in the document]
Comment (RBBentley) [reply core/frontend/src/tile/RealityTile.ts:382]: Good point - I ended up with 20Km empirically -- trying to come up with a value that would not produce too many stepchildren.   I would prefer to use a smaller value but even 2000M causes excessive subdivision and thousands of tiles (I tried it).

[the remark this answers is not in the document]
Comment (RBBentley) [reply core/frontend/src/tile/RealityTile.ts:382]: FWIW -- The tile radius is grossly exaggerated as it is calculated from the range diagonal in a bounding box aligned to the world.

File: core/frontend/src/tile/RealityTile.ts [hunk 3]
```diff
@@ -275,4 +311,124 @@ export class RealityTile extends Tile {
...
+    return this._graphic;
+  }
+  public override disposeContents(): void {
+    super.disposeContents();
+    this._reprojectedGraphic = dispose(this._reprojectedGraphic);
+  }
+}
+
+/** When additive refinement is used (as in the Cesium OSM tileset) it is not possible to accurately reproject very large, low level tiles
+ * In this case we create addtitional "step" children (grandchildren etc. ) that will clipped portions display the their ancestor's additive geometry.
```
[the remark this answers is not in the document]
Comment (RBBentley) [reply core/frontend/src/tile/RealityTile.ts]: Fixed, Thanks.

File: core/frontend/src/tile/RealityTile.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (AlainRobertAtBentley) [reply core/frontend/src/tile/RealityTile.ts:382]: Okay, We will try it out with 20km and see how it performs.

**Comment by AlainRobertAtBentley** [comment]
I had Richard review (he knows more about tiles and ECEF and transform than I do) and it said it looked good to him. The GeoConverter stuff appears correct too.

**Review by AlainRobertAtBentley — APPROVED** [review]

**Comment by pmconne** [comment]
I opened the TPB iModel. Turned buildings on and background map off. Panned slightly left from initial view then zoomed out a couple times. The app froze with 6 active tiles loading. I killed it after ~2 minutes.

**Comment by pmconne** [comment]
After keying in `fdt set building display`, frustum does not appear to get expanded until a view navigation occurs. Until then, buildings are clipped by previous near/far planes.

**Review by pmconne — CHANGES_REQUESTED** [review]
Please investigate OOM crashes.

I opened the TPB iModel. Turned buildings on and background map off. Panned slightly left from initial view then zoomed out a couple times. The app froze with 6 active tiles loading. I killed it after ~2 minutes.
Reproduced this in similar manner again. Application window went blank white after ~30 seconds (render process gone).
Reproduced a third time (disabling map is not required). Memory usage reported by Task Manager rapidly skyrockets past 4 GB before the render process dies.

**Comment by RBBentley** [comment]
> Please investigate OOM crashes.
> 
> I opened the TPB iModel. Turned buildings on and background map off. Panned slightly left from initial view then zoomed out a couple times. The app froze with 6 active tiles loading. I killed it after ~2 minutes.
> Reproduced this in similar manner again. Application window went blank white after ~30 seconds (render process gone).
> Reproduced a third time (disabling map is not required). Memory usage reported by Task Manager rapidly skyrockets past 4 GB before the render process dies.

Yes - Looking at this now.  For some reason there is a tile where the subdivision never finishes.

**Comment by RBBentley** [comment]
> Please investigate OOM crashes.
> 
> I opened the TPB iModel. Turned buildings on and background map off. Panned slightly left from initial view then zoomed out a couple times. The app froze with 6 active tiles loading. I killed it after ~2 minutes.
> Reproduced this in similar manner again. Application window went blank white after ~30 seconds (render process gone).
> Reproduced a third time (disabling map is not required). Memory usage reported by Task Manager rapidly skyrockets past 4 GB before the render process dies.

Fixed.   I was stupidly testing radius for the subdivision termination which is never reduced when you quadrant the parent.

**Comment by RBBentley** [comment]
@AlainRobertAtBentley  - After fixing the subdivision criteria where I was stupidly testing the tile radius, I was able to reduce the termination threshold to 2000M.   This reduces the potential altitude error to less than a meter.

**Comment by pmconne** [comment]
Is 335 tiles an appropriate number to be selected when viewing this region?
image

Zoomed out with scene frozen:
image

**Comment by RBBentley** [comment]
> Is 335 tiles an appropriate number to be selected when viewing this region?
> image
> 
> Zoomed out with scene frozen:
> image

Yes - Unfortunately this is.  In this region we have additive tiles contributing to the view from depth 3 all the way down to 14.  I have to include stepchildren for each.  It doesn't look quite as overly dense if you look at the reality range boxes "fdt reality bounds" which displays the boxes from the range corners- which are oriented more reasonably.

image

**Comment by pmconne** [comment]
This top view of Exton is drawing over 3000 building tiles. FPS < 10 when stationary. It drops below 5 when moving the mouse cursor.
image

**Comment by RBBentley** [comment]
> This top view of Exton is drawing over 3000 building tiles. FPS < 10 when stationary. It drops below 5 when moving the mouse cursor.
> image

Yes - I see this.  I'm going to have to come up with a better criteria for ending the subdivision.

**Comment by RBBentley** [comment]
> > This top view of Exton is drawing over 3000 building tiles. FPS < 10 when stationary. It drops below 5 when moving the mouse cursor.
> > image
> 
> Yes - I see this. I'm going to have to come up with a better criteria for ending the subdivision.

Fix pushed - Please retest...

**Comment by pmconne** [comment]
> Fix pushed - Please retest...

Significantly improved.

**Review by pmconne — APPROVED** [review]

**Comment by pmconne** [comment]
@Mergifyio update

**Comment by pmconne** [comment]
@Mergifyio backport release/2.19.x

**Pull request merged**